File: controllers/src/main/java/com/example/controllers/comicdetail/ComicDetailViewModelDelegate.java
package com.example.controllers.comicdetail;

import android.util.Log;
import android.view.View;

import com.example.controllers.commons.CardClickListener;
import com.example.controllers.commons.ProcessedMarvelCharacter;
import com.example.controllers.commons.ProcessedMarvelComic;
import com.example.controllers.commons.ProcessedMarvelEvent;
import com.example.controllers.commons.ProcessedMarvelItemBase;
import com.example.controllers.commons.ProcessedMarvelSeries;
import com.example.controllers.commons.ProcessedMarvelStory;
import com.example.mviframework.BaseMviDelegate;
import com.example.mviframework.Change;
import com.example.mviframework.Reducer;
import com.example.mviframework.Runner;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;

public class ComicDetailViewModelDelegate extends BaseMviDelegate<State, ComicDetailViewModelDelegate.InnerState, Effect> {
    ComicDetailNetworkInterface networkInterface;
    ProcessedMarvelComic comic;

    CardClickListener cardClickListener = new CardClickListener<ProcessedMarvelItemBase>() {

        @Override
        public void invoke(View view, ProcessedMarvelItemBase item) {
            enqueue(new Reducer<InnerState, Effect>() {
                @Override
                public Change<InnerState, Effect> reduce(InnerState innerState) {
                    return withEffects(innerState, new Effect.ClickCardEffect<ProcessedMarvelItemBase>(view, item));
                }
            });
        }
    };

    Runner callbackRunner = dispatch(new Reducer<InnerState, Effect>() {
        @Override
        public Change<InnerState, Effect> reduce(InnerState innerState) {
            InnerState newInnerState = innerState.copy();
            newInnerState.loading = false;
            newInnerState.error = false;
            return withEffects(newInnerState, new Effect.ImageLoaded());
        }
    });

    public ComicDetailViewModelDelegate(ComicDetailNetworkInterface networkInterface, ProcessedMarvelComic comic) {
        this.networkInterface = networkInterface;
        this.comic = comic;
        loadCharacters(comic.id);
        loadSeries(comic.id);
        loadStories(comic.id);
        loadEvents(comic.id);

    }

    private void loadEvents(int id) {
        if(networkInterface==null) {
            return;
        }
        Observable<List<ProcessedMarvelEvent>> observable = networkInterface
                .loadEventsForComic(comic.id)
                .onErrorReturn(throwable -> {
                    return new ArrayList<>();
                });
        Observable<Reducer<InnerState, Effect>> reducerObservable = observable
                .map( events -> {
                    return new Reducer<InnerState, Effect>() {
                        @Override
                        public Change<InnerState, Effect> reduce(InnerState innerState) {
                            InnerState newInnerState = innerState.copy();
                            newInnerState.events = events;
                            return asChange(newInnerState);
                        }
                    };
                });
        enqueue(reducerObservable);
    }

    private void loadStories(int id) {
        if(networkInterface==null) {
            return;
        }
        Observable<List<ProcessedMarvelStory>> observable = networkInterface
                .loadStoriesForComic(comic.id)
                .onErrorReturn(throwable -> {
                    return new ArrayList<>();
                });
        Observable<Reducer<InnerState, Effect>> reducerObservable = observable
                .map( stories -> {
                    return new Reducer<InnerState, Effect>() {
                        @Override
                        public Change<InnerState, Effect> reduce(InnerState innerState) {
                            InnerState newInnerState = innerState.copy();
                            newInnerState.stories = stories;
                            return asChange(newInnerState);
                        }
                    };
                });
        enqueue(reducerObservable);
    }

    private void loadSeries(int id) {
        if (networkInterface == null) {
            return;
        }
        Observable<List<ProcessedMarvelSeries>> observable = networkInterface
                .loadSeriesForComic(comic.id)
                .onErrorReturn(throwable -> {
                    return new ArrayList<>();
                });
        Observable<Reducer<InnerState, Effect>> reducerObservable = observable
                .map(series -> {
                   // Log.d("Vartika3", "Reducer1: series "+series);
                    return new Reducer<InnerState, Effect>() {

                        @Override
                        public Change<InnerState, Effect> reduce(InnerState innerState) {
                            //Log.d("Vartika3", "Reducer2: series "+series);
                            InnerState newInnerState = innerState.copy();
                            newInnerState.series = series;
                            return asChange(newInnerState);
                        }
                    };
                });
        enqueue(reducerObservable);
    }

    private void loadCharacters(int id) {
        if (networkInterface == null) {
            return;
        }
        Observable<List<ProcessedMarvelCharacter>> observable = networkInterface
                .loadCharactersForComic(comic.id)
                .onErrorReturn(throwable -> {
                    return new ArrayList<>();
                });
        Observable<Reducer<InnerState, Effect>> reducerObservable = observable
                .map(characters -> {
                    // Log.d("Vartika3", "Reducer1: series "+series);
                    return new Reducer<InnerState, Effect>() {

                        @Override
                        public Change<InnerState, Effect> reduce(InnerState innerState) {
                            //Log.d("Vartika3", "Reducer2: series "+series);
                            InnerState newInnerState = innerState.copy();
                            newInnerState.characters = characters;
                            return asChange(newInnerState);
                        }
                    };
                });
        enqueue(reducerObservable);
    }

    @Override
    public Change<InnerState, Effect> getInitialChange() {
        return asChange(new InnerState(comic, null, null, null, null, true, false));
    }

    @Override
    public State mapState(ComicDetailViewModelDelegate.InnerState innerState) {
        State state = new State(innerState.comic, innerState.characters, innerState.series, innerState.stories, innerState.events,  innerState.loading, innerState.error, callbackRunner, cardClickListener);
        Log.d("Vartika", "Comic Detail mapState: "+state.toString());
        return state;
    }
    static class InnerState {
        boolean loading;
        boolean error;
        ProcessedMarvelComic comic;
        List<ProcessedMarvelCharacter> characters;
        List<ProcessedMarvelSeries> series;
        List<ProcessedMarvelStory> stories;
        List<ProcessedMarvelEvent> events;

        InnerState(ProcessedMarvelComic comic, List<ProcessedMarvelCharacter> characters, List<ProcessedMarvelSeries> series, List<ProcessedMarvelStory> stories, List<ProcessedMarvelEvent> events, boolean loading, boolean error) {
            this.comic = comic;
            this.loading = loading;
            this.error = error;
            this.characters = characters;
            this.series = series;
            this.stories = stories;
            this.events = events;
        }
        InnerState copy() {
            return new InnerState(comic, characters, series, stories, events, loading, error);
        }
    }
}

File: controllers/src/main/java/com/example/controllers/comicdetail/ComicDetailViewModelDelegate.kt
package com.example.controllers.comicdetail

import android.util.Log
import android.view.View
import com.example.controllers.commons.*
import com.example.mviframework.BaseMviDelegate
import com.example.mviframework.Change
import com.example.mviframework.Reducer
import io.reactivex.Observable
import java.util.*

class ComicDetailViewModelDelegate(val networkInterface: ComicDetailNetworkInterface, val comic: ProcessedMarvelComic) : BaseMviDelegate<State, ComicDetailViewModelDelegate.InnerState, Effect>() {
    var cardClickListener: CardClickListener<*> = object : CardClickListener<ProcessedMarvelItemBase> {
        override fun invoke(view: View, item: ProcessedMarvelItemBase) {
            enqueue { innerState -> withEffects(innerState, Effect.ClickCardEffect(view, item)) }
        }
    }
    var callbackRunner = dispatch { innerState ->
        val newInnerState = innerState.copy()
        newInnerState.loading = false
        newInnerState.error = false
        withEffects(newInnerState, Effect.ImageLoaded<Any?>())
    }
    fun onUrlLinkClicked(url: String) {
        enqueue {
            withEffects(it, Effect.OpenUrl(url))
        }
    }

    private fun loadEvents(id: Int) {
        if (networkInterface == null) {
            return
        }
        val observable = networkInterface
                .loadEventsForComic(comic.id)
                .onErrorReturn { throwable: Throwable -> ArrayList() }
        val reducerObservable: Observable<Reducer<InnerState, Effect>> = observable
                .map { events: List<ProcessedMarvelEvent> ->
                    Reducer { innerState ->
                        val newInnerState = innerState.copy()
                        newInnerState.events = events
                        asChange(newInnerState)
                    }
                }
        enqueue(reducerObservable)
    }

    private fun loadStories(id: Int) {
        if (networkInterface == null) {
            return
        }
        val observable = networkInterface
                .loadStoriesForComic(comic.id)
                .onErrorReturn { throwable: Throwable -> ArrayList() }
        val reducerObservable: Observable<Reducer<InnerState, Effect>> = observable
                .map { stories: List<ProcessedMarvelStory> ->
                    Reducer { innerState ->
                        val newInnerState = innerState.copy()
                        newInnerState.stories = stories
                        asChange(newInnerState)
                    }
                }
        enqueue(reducerObservable)
    }

    private fun loadSeries(id: Int) {
        if (networkInterface == null) {
            return
        }
        val observable = networkInterface
                .loadSeriesForComic(comic.id)
                .onErrorReturn { throwable: Throwable -> ArrayList() }
        val reducerObservable: Observable<Reducer<InnerState, Effect>> = observable
                .map { series: List<ProcessedMarvelSeries> ->
                    Reducer { innerState -> //Log.d("Vartika3", "Reducer2: series "+series);
                        val newInnerState = innerState.copy()
                        newInnerState.series = series
                        asChange(newInnerState)
                    }
                }
        enqueue(reducerObservable)
    }

    private fun loadCharacters(id: Int) {
        if (networkInterface == null) {
            return
        }
        val observable = networkInterface
                .loadCharactersForComic(comic.id)
                .onErrorReturn { throwable: Throwable -> ArrayList() }
        val reducerObservable: Observable<Reducer<InnerState, Effect>> = observable
                .map { characters: List<ProcessedMarvelCharacter> ->
                    Reducer { innerState -> //Log.d("Vartika3", "Reducer2: series "+series);
                        val newInnerState = innerState.copy()
                        newInnerState.characters = characters
                        asChange(newInnerState)
                    }
                }
        enqueue(reducerObservable)
    }

    override fun getInitialChange(): Change<InnerState, Effect> {
        return asChange(InnerState(comic, null, null, null, null, null, true, false))
    }

    override fun mapState(innerState: InnerState): State {
        val state = State(
                innerState.comic,
                innerState.comic.urls.map {
                    ProcessedURLItem(it.type, {this::onUrlLinkClicked.invoke(it.url)});
                },
                innerState.characters,
                innerState.series,
                innerState.stories,
                innerState.events,
                innerState.loading,
                innerState.error,
                callbackRunner,
                cardClickListener)
        Log.d("Vartika", "Comic Detail mapState: $state")
        return state
    }

    class InnerState(var comic: ProcessedMarvelComic, var urls: List<ProcessedURLItem>?, var characters: List<ProcessedMarvelCharacter>?, var series: List<ProcessedMarvelSeries>?, var stories: List<ProcessedMarvelStory>?, var events: List<ProcessedMarvelEvent>?, var loading: Boolean, var error: Boolean) {
        fun copy(): InnerState {
            return InnerState(comic, urls, characters, series, stories, events, loading, error)
        }
    }

    init {
        loadCharacters(comic.id)
        loadSeries(comic.id)
        loadStories(comic.id)
        loadEvents(comic.id)
    }
}